File: Testing_Catalog/PythonAlgorithms/Sort/quickSort.py
'''
Quick Sort: Python

Runtime Complexity within O(n*Log2(n))

Takes array, returns sorted array
'''

def swap(array: list, i: int, j: int):
    temp = array[i]
    array[i] = array[j]
    array[j] = temp
    return

def partition(array: list, low: int, high: int) -> int:
    i = low
    j = high

    while i < j:
        while (i < high) and (array[i] <= array[low]):
            i += 1
        while array[j] > array[low]:
            j -= 1
        if i < j:
            swap(array, i, j)
    swap(array, low, j)

    return j

def quick_sort(array: list, low: int, high: int):
    if low >= high:
        return

    pivot = partition(array, low, high)

    quick_sort(array, low, pivot - 1)
    quick_sort(array, pivot + 1, high)

    return

File: Testing_Catalog/CAlgorithms/Sort/quickSort.c
//================================================
//================================================
// Quick Sort Algorigthm
// quickSort.c
//
// Time Complexity: O(n*Log2(n))
// Takes array, returns sorted array
//================================================
//================================================
#include "main.h"

// utility
int partition(int* array, int low, int high)
{
    //parameters
    int pivot = low;
    int i = low;
    int j = high;

    //iterate and swap untill indices cross
    while (i < j) {
        while ((array[i] <= array[pivot]) && (i < high))
            ++i;
        while (array[j] > array[pivot])
            --j;
        if (i < j)
            swap(array, i, j);
    }
    swap(array, pivot, j);

    return j;
}

// algorithm
void quick_sort(int* array, int low, int high)
{
    if (low >= high)
        return;

    int pivot = partition(array, low, high);

    quick_sort(array, low, pivot - 1);
    quick_sort(array, pivot + 1, high);

    return;
}

